source/features/preserve-file-finder-term.tsx

```diff
@@ -0,0 +1,120 @@
+const initializeValues = async (): Promise<void> => {
+	const valuesFromStorage = (await browser.storage.local.get({fileFinderMemory: '', fileFinderBuffer: ''}));
+	fileFinderMemory = valuesFromStorage.fileFinderMemory;
+	fileFinderBuffer = valuesFromStorage.fileFinderBuffer;
+};
+
+const keyDownHandler = ({key, target}: KeyboardEvent): void => {
+	const nodeName = (target as Element).nodeName;
+	if (nodeName !== 'INPUT' && nodeName !== 'TEXTAREA') {
+		if (flag) {
+			const input = select<HTMLInputElement>(FILE_FINDER_INPUT_SELECTOR);
+			if (!input && isValidCharacter(key)) {
+				setBuffer(fileFinderBuffer.concat(key));
+			}
+		} else if (key === 't') {
+			flag = true;
+		}
+	}
+};
+
+// The current search term is stored in local storage
+const setText = async (value: string): Promise<void> => {
+	fileFinderMemory = value;
+	await browser.storage.local.set({fileFinderMemory});
+};
+
+// Chars pressed after 't' is stored in local storage
+const setBuffer = async (value: string): Promise<void> => {
+	fileFinderBuffer = value;
+	await browser.storage.local.set({fileFinderBuffer});
+};
+
+const inputHandler = async (event: DelegateEvent<InputEvent, HTMLInputElement>): Promise<void> => {
+	const value: string = (event.delegateTarget).value;
+	await setText(value);
+};
+
+// Set the input field value & trigger event
+const setValueInField = (value: string): void => {
+	const inputElement = select<HTMLInputElement>(FILE_FINDER_INPUT_SELECTOR);
+	if (inputElement) {
+		inputElement.value = value;
+		inputElement.dispatchEvent(new Event('input')); // Manually trigger event to trigger search
+	}
+};
+
+// Function that will set the current value of input field to buffer or memory
+const setValueFromBufferOrText = async (): Promise<void> => {
+	if (fileFinderBuffer.length > 0) {
+		setValueInField(fileFinderBuffer);
+		await setText(fileFinderBuffer);
+		await setBuffer('');
+	} else if (fileFinderMemory) {
+		setValueInField(fileFinderMemory);
+	}
+};
+
+// Function that is used for polling until file finder input becomes ready
+function pollFn(): boolean | null {
+	const inputElement = select<HTMLInputElement>(FILE_FINDER_INPUT_SELECTOR);
+	if (inputElement) {
+		setValueFromBufferOrText();
+		flag = false;
+		return true;
+	}
+
+	return null;
+}
+
+async function init(): Promise<void> {
+	await initializeValues();
+	if (features.isFileFinder()) {
+		await setValueFromBufferOrText();
+		delegate(FILE_FINDER_INPUT_SELECTOR, 'input', inputHandler);
+	} else { // Wait for file finder page to be opened
+		poll(pollFn, 300);
+	}
+
+	window.addEventListener('keydown', keyDownHandler);
+}
+
+function deinit(): void {
+	window.removeEventListener('keydown', keyDownHandler);
+}
+
+features.add({
+	id: __featureName__,
+	description: 'Preserve search term in file finder',
+	screenshot: false,
+	include: [
+		features.isRepo
+	],
+	load: features.onAjaxedPages,
+	init,
+	deinit
+});
```

Comment:
# Pre-load buffer

The buffer only makes sense between the time the user starts navigating to FF (file finder) and the time it loads. This period can be detected by listening to GitHub’s PJAX events. I’m not sure, but maybe their “pjax start” event also includes which page is about to load, so this period can be detected *exactly.*

Once this buffer is added to the field, it should be discarded immediately.

This feature can only work if FF only loads via ajax, every time. If it doesn’t, this feature isn’t feasible: we’d lose key presses between the time the page unloads and our content scripts load on the new page. 

Because of this, the buffer can be stored in a regular `let` variable.

Reply:
In my tests, the FF is loaded via PJAX only when navigated from Code, Issues & Pull Requests tab. Initially I was using a simple `let` variable just like you said. Then I went with persisting the buffer in storage to store key presses when the file finder is not loaded via PJAX.  
Another issue I stumbled across was key presses coinciding with other shortcuts, and the FF not opening up.

TL;DR: FF is not loaded via ajax every time. Should we drop this feature, and just continue with text restoration in FF?

Reply:
Code, issues and PRs covers most situations. I think it’s worth implementing it in these cases. This implementation should be easier than before since you can just read write synchronously